src/symbol/place_symbols.js

```diff
@@ -0,0 +1,260 @@
+// @flow
+
+const symbolSize = require('./symbol_size');
+
+import type SymbolBucket, {SymbolInstance} from '../data/bucket/symbol_bucket';
+import type OpacityState from './opacity_state';
+import type CollisionIndex from './collision_index';
+import type CollisionBoxArray from './collision_box';
+import type {StructArray} from '../util/struct_array';
+const mat4 = require('@mapbox/gl-matrix').mat4;
+
+module.exports = {
+    updateOpacities: updateOpacities,
+    place: place
+};
+
+function updateOpacity(symbolInstance: SymbolInstance, opacityState: OpacityState, targetOpacity: number, opacityUpdateTime: number, collisionFadeTimes: any): boolean {
+    const initialVisible = opacityState.opacity !== 0;
+    const initialTargetVisible = opacityState.targetOpacity !== 0;
+    if (symbolInstance.isDuplicate) {
+        opacityState.opacity = 0;
+        opacityState.targetOpacity = 0;
+    } else {
+        if (opacityState.targetOpacity !== targetOpacity) {
+            collisionFadeTimes.latestStart = opacityUpdateTime;
+        }
+        const increment = collisionFadeTimes.duration ? ((opacityUpdateTime - opacityState.time) / collisionFadeTimes.duration) : 1;
+        opacityState.opacity = Math.max(0, Math.min(1, opacityState.opacity + (opacityState.targetOpacity === 1 ? increment : -increment)));
+        opacityState.targetOpacity = targetOpacity;
+        opacityState.time = opacityUpdateTime;
+    }
+    return (!initialVisible && !initialTargetVisible && opacityState.targetOpacity !== 0) ||
+           ((initialVisible || initialTargetVisible) && (opacityState.opacity === 0 && opacityState.targetOpacity === 0));
+}
+
+const shift25 = Math.pow(2, 25);
+const shift24 = Math.pow(2, 24);
+const shift17 = Math.pow(2, 17);
+const shift16 = Math.pow(2, 16);
+const shift9 = Math.pow(2, 9);
+const shift8 = Math.pow(2, 8);
+const shift1 = Math.pow(2, 1);
+
+// All four vertices for a glyph will have the same opacity state
+// So we pack the opacity into a uint8, and then repeat it four times
+// to make a single uint32 that we can upload for each glyph in the
+// label.
+function packOpacity(opacityState: OpacityState): number {
+    if (opacityState.opacity === 0 && opacityState.targetOpacity === 0) {
+        return 0;
+    } else if (opacityState.opacity === 1 && opacityState.targetOpacity === 1) {
+        return 4294967295;
+    }
+    const targetBit = opacityState.targetOpacity === 1 ? 1 : 0;
+    const opacityBits = Math.floor(opacityState.opacity * 127);
+    return opacityBits * shift25 + targetBit * shift24 +
+        opacityBits * shift17 + targetBit * shift16 +
+        opacityBits * shift9 + targetBit * shift8 +
+        opacityBits * shift1 + targetBit;
+}
```

Comment:
what's the advantage of the multiplication instead of using instead of the bitshift operator?

Reply:
The problem was that the high bits could be interpreted as negative numbers, and that threw off the addition logic.